### DISCUSSION TITLE
How to exclude absolute paths?

### DISCUSSION THREAD

**Discussion by rieje** [BurntSushi/ripgrep#2421] [Q&A]
How to exclude absolute paths with ripgrep? `--glob` and `--ignore-file` is based on gitignore file which works on relative paths only. So is fd's `--ignore-file`, yet absolute paths in ignore files used by fd appears to work (was suggested as a workaround to fd's `--exclude` which doesn't work with absolute paths).

**Comment by BurntSushi** [comment]
Please make your question concrete by providing specific examples that others can reproduce. Does something not work? What did you try? What is actual output and what is expected output?

**Reply by rieje** [reply]
Want to exclude /home/rieje/repo from results.

Both attempts show results of files from that directory:

    rg --no-config --glob '!/home/rieje/repo' mystring
    # where ~/.config/rg/general.rgignore contains /home/rieje/repo 
    rg --no-config -ignore-file ~/.config/rg/general.rgignore mystring

Expected result: files from /home/rieje/repo are excluded from results. Also tried with `**` appended to glob pattern.

**Reply by BurntSushi** [reply]
Yeah that isn't going to work because globs match file paths. You're searching the current working directory, so paths look like `./repo/foo/bar`. Therefore, `/home/reije/repo` doesn't match that. You probably want to do `echo repo > /home/rieje/.ignore`.

**Reply by tmccombs** [reply]
This is something that has come up several times with `fd` as well, for example https://github.com/sharkdp/fd/issues/1337 .

One issue with following the pattern of gitignore, is that with git, it makes since for "absolute" patterns in the ignore rules to be relative to the git repo, because the repo that is the universe of git commands. But for `fd` and `ripgrep` there isn't really an equivalent. The `ignore` crate treats patterns that start with `/` as relative to the working directory, but I think that is pretty surprising behavior to most users. If I were to design it today, I would have made patterns that start with `/` match against the absolute path, and maybe have special support for `./` to match relative to the current working directory. 

Unfortunately, changing that now would be backwards incompatible, but maybe when the ignore rewrite finally happens, such a breaking change would be acceptable? Or maybe there could be some other way to indicate a pattern is absolute? Perhaps if it starts with ":" or "@"?

Edit:

I'd also like to point out that if you are in a git repo that has a folder "a", and you run `git ls-files -o --exclude /foo` from the "a" directory, it won't exclude `/foo`, you instead need to run `git ls-files -o --exclude /a/foo`.  So treating / as relative to the current directory doesn't even match the behavior of git.

**Reply by tmccombs** [reply]
Oh, and to be clear, I think it makes since for patterns starting with "/" to be relative to the directory of the file for .ignore  or .rgignore files.  I think matching absolute paths only makes since for  patterns in an `--ignore-file` file or passed in with the `--glob` option.

**Reply by BurntSushi** [reply]
Yeah I agree there is something to figure out here. I basically think that `ignore` filtering needs to be re-thought from first principles while considering the outstanding bugs around this sort of thing. It's on my list of things to do, but it's not a high priority for me unfortunately.
